### PR TITLE
empirical disp updates

### LINKED ISSUE #1710: DFTD4 interface in Psi4
Dear Psi4 Developers,

Is there any plan to add the newly DFTD4 interface in Psi4

### PR DESCRIPTION
**Pull request by loriab**
## Todos

- [x] switch gcp interface to QCEngine
- [x] add dftd4 interface calling QCEngine
- [x] update types return for disp. engine can handle ndarray, so leave arrays as np and shaped rather than flat lists
- [x] more tests -- psiapi, parameters extend func
- [x] add docs
- [x] note min qcng version, prob. v0.19
- [x] this passes cleanly locally but that's with custom engine and dftd4, so several PRs ahead of this one
- [x] allow doi as citations for dft
- [x] note that for `Mol.run_dftd4`, func overrides parameters in keeping with dftd4 API behavior, whereas in dftd3, parameters extend or override func.

## Questions
- [x] @jeffschriber should fisapt grab 2-body disp analysis or total disp analysis
- [x] a couple of the interface updates should be in v1.4 but d4 itself can be in v1.5. so this can get split after upstream settles down.

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Aug 2021 Notes
- [x] rebased. the only thing that's left of the original PR checklist is dftd4 itself. gcp and dftd3 updates were pulled in in #2180.
- [x] note that it's not the dftd4 exe that this (that is, qcng) needs; rather, it's the dftd4 pymod that's needed.
- [x] one _can_ use the c-f `dftd4-python` conda package if you know how to set up your env to install both it and psi4 deps. But for linux only, I've prepared a `dftd4` conda package (not in final build form) off `-c psi4/label/dev`.

## Status
- [x] Ready for review
- [x] Ready for merge

EDIT: closes #1710

### CODE REVIEW

File: psi4/driver/procrouting/empirical_dispersion.py [hunk 1]
```diff
@@ -36,12 +36,12 @@
 from psi4.driver import p4util
 from psi4.driver import driver_findif
 from psi4.driver.p4util.exceptions import ValidationError
-from psi4.driver.qcdb import interface_gcp as gcp
 
 _engine_can_do = collections.OrderedDict([('libdisp', ['d1', 'd2', 'chg', 'das2009', 'das2010']),
                                           ('dftd3', ['d2', 'd3zero', 'd3bj', 'd3mzero', 'd3mbj']),
                                           ('nl', ['nl']),
                                           ('mp2d', ['dmp2']),
+                                          ("dftd4", ["d4bj"]),
```
Comment (PeterKraus) [inline psi4/driver/procrouting/empirical_dispersion.py:44]: Single quotes. This applies also below.

File: psi4/driver/procrouting/empirical_dispersion.py [hunk 2]
```diff
@@ -201,7 +201,7 @@ def compute_energy(self, molecule: 'psi4.core.Molecule', wfn: 'psi4.core.Wavefun
                         'level_hint': self.dashlevel,
                         'params_tweaks': self.dashparams,
                         'dashcoeff_supplement': self.dashcoeff_supplement,
-                        'save_pairwise_dispersion': self.save_pairwise_disp,
+                        'pair_resolved': self.save_pairwise_disp,
```
Comment (PeterKraus) [inline psi4/driver/procrouting/empirical_dispersion.py:204]: Why is this renamed?

Comment (loriab) [reply psi4/driver/procrouting/empirical_dispersion.py:204]: dftd4 includes a pairwise analysis option directly in its qcschema API under the name `pair_resolved`. since the `save_pairwise_dispersion` option we invented to trigger `-anal` for dftd3 has only been in a couple months, it seemed early enough to drop the d3 name and conform to d4.

File: psi4/driver/procrouting/empirical_dispersion.py (hunk 1 above)
Comment (loriab) [reply psi4/driver/procrouting/empirical_dispersion.py:44]: while I have long preferred `'` and `"""`, I'm expecting `black` will someday steamroll through and impose `"`, so I'm averting lines changed in future.

File: doc/sphinxman/source/dftd3.rst [hunk 3]
```diff
@@ -199,13 +199,15 @@ B3LYP, a B3LYP-D2, and a B3LYP-D3 (zero-damping) energy. ::
...
     energy('b3lyp-d3')
+    energy('b3lyp-d4')
 
 Consult the table :ref:`-D Functionals <table:dft_disp>` to see for each
 functional what corrections are available and what default parameters
 define them. The dispersion correction is available after a calculation in
 the PSI variable :psivar:`DISPERSION CORRECTION ENERGY`.
 By default, the output from the ``dftd3``
 program is suppressed; to see it in the output file, set print > 2.
+No next output is available from the ``dftd4`` program.
```
Comment (JonathonMisiewicz) [inline doc/sphinxman/source/dftd3.rst:210]: Do you mean no _additional_ output?

File: psi4/driver/qcdb/molecule.py [hunk 4]
```diff
@@ -1336,6 +1336,96 @@ def run_dftd3(self, func: str = None, dashlvl: str = None, dashparam: Dict = Non
...
+        func : str, optional
+            Name of functional (func only, func & disp, or disp only) for
+            which to compute dispersion (e.g., blyp, BLYP-D2, blyp-d3bj,
+            blyp-d3(bj), hf+d). Any or all parameters initialized
+            from `dashcoeff[dashlvl][func]` can be overwritten via
+            `dashparam`.
+        dashlvl : str, optional
+            Name of dispersion correction to be applied (e.g., d, D2,
+            d3(bj), das2010). Must be key in `dashcoeff` or "alias" or
+            "formal" to one.
```
Comment (JonathonMisiewicz) [inline psi4/driver/qcdb/molecule.py:1353]: Do you mean to _run_?

File: psi4/driver/qcdb/molecule.py [hunk 5]
```diff
@@ -1336,6 +1336,96 @@ def run_dftd3(self, func: str = None, dashlvl: str = None, dashparam: Dict = Non
         elif derint == 1:
             return jobrec['extras']['qcvars']['DISPERSION CORRECTION GRADIENT']
 
+    def run_dftd4(self, func=None, dashlvl=None, dashparam=None, dertype=None, verbose=1):
+        """Compute dispersion correction via Grimme's DFTD4 program.
+
+        Parameters
+        ----------
+        func : str, optional
+            Name of functional (func only, func & disp, or disp only) for
```
Comment (JonathonMisiewicz) [inline psi4/driver/qcdb/molecule.py:1345]: What if this isn't supplied?

**Review by JonathonMisiewicz — APPROVED** [review]
A few documentation comments, but LGTM otherwise.

**Comment by JonathonMisiewicz** [comment]
Also, please `grep` for comments related to `dftd3` in the code and update them as needed. `proc.py::scf_helper` certainly needs to be updated.

**Review by hokru — APPROVED** [review]

File: doc/sphinxman/source/dftd3.rst [hunk 6]
```diff
@@ -172,21 +172,21 @@ All parameters characterizing the dispersion correction are taken from
...
+Running DFTD3 or DFTD4
+~~~~~~~~~~~~~~~~~~~~~~
 
 A number of *a posteriori* dispersion corrections are available in
 |PSIfour|.  While some are computed within |PSIfours| codebase (-D1, -D2,
--CHG, -DAS2009, -DAS2010), the -D3 correction and its variants are
-available only through the ``DFTD3`` program.  Once installed, the
+-CHG, -DAS2009, -DAS2010), the -D3 or -D4 corrections and their variants are
+available only through the ``DFTD3`` or ``DFTD4`` programs. Once installed, the
 ``dftd3``/|PSIfour| interface is transparent, and all corrections are
```
Comment (PeterKraus) [inline doc/sphinxman/source/dftd3.rst:182]: dftd3/psi4 as well as dftd4/psi4 interfaces, presumably.

File: psi4/driver/procrouting/empirical_dispersion.py [hunk 7]
```diff
@@ -42,6 +42,7 @@
                                           ('dftd3', ['d2', 'd3zero', 'd3bj', 'd3mzero', 'd3mbj']),
                                           ('nl', ['nl']),
                                           ('mp2d', ['dmp2']),
+                                          ("dftd4", ["d4bjeeqatm"]),
```
Comment (PeterKraus) [inline psi4/driver/procrouting/empirical_dispersion.py:45]: Double quotes here, but not above. Also on line 60 and throughout this file.

File: psi4/driver/qcdb/molecule.py [hunk 8]
```diff
@@ -1336,6 +1336,96 @@ def run_dftd3(self, func: str = None, dashlvl: str = None, dashparam: Dict = Non
...
+        gradient : ndarray
+            When `dertype=1`, (nat, 3) gradient [Eh/a0].
+        (energy, gradient) : tuple of float and ndarray
+            When `dertype=None`, both energy [Eh] and (nat, 3) gradient [Eh/a0].
+
+        Notes
+        -----
+        This function wraps the QCEngine dftd4 harness which wraps the internal DFTD4 Python API.
+        As such, the upstream convention of `func` trumping `dashparam` holds, rather than the
+        :py:func:`run_dftd3` behavior of `dashparam` extending or overriding `func`.
```
Comment (PeterKraus) [inline psi4/driver/qcdb/molecule.py:1378]: Is there any way to harmonise these two? The way I understand it, we do the same thing Psi4-side, but the behaviour of dftd3 and dftd4 differs when both func and dashparams are provided.

In any case, not Psi4's but rather qcng's problem...

I'm guessing the only problem might be when one wants to run, say, B3LYP (for which `func` values are defined) with a custom set of parameters - is this possible to do with dftd4?

File: psi4/driver/qcdb/molecule.py [hunk 9]
```diff
@@ -1336,6 +1336,96 @@ def run_dftd3(self, func: str = None, dashlvl: str = None, dashparam: Dict = Non
...
+            blyp-d3(bj), hf+d). Any or all parameters initialized
+            from `dashcoeff[dashlvl][func]` can be overwritten via
+            `dashparam`.
+        dashlvl : str, optional
+            Name of dispersion correction to be applied (e.g., d, D2,
+            d3(bj), das2010). Must be key in `dashcoeff` or "alias" or
+            "formal" to run.
+        dashparam : dict, optional
+            Values for the same keys as `dashcoeff[dashlvl]['default']`
+            used to override any or all values initialized by `func`.
```
Comment (PeterKraus) [inline psi4/driver/qcdb/molecule.py:1356]: Given your comment below, this is not true for dftd4.

**Review by PeterKraus — CHANGES_REQUESTED** [review]
This might be a good place to add at least the most common -D4 functionals to `test_dft_benchmark.py`.

File: psi4/driver/qcdb/molecule.py (hunk 8 above)
Comment (awvwgk) [reply psi4/driver/qcdb/molecule.py:1378]: See related discussion here: https://github.com/dftd4/dftd4/issues/98

Short summary:

`dftd4` does not provide the possibility to select a functional, say B3LYP and only change the *s8*, this is by design. Either the downstream user hands the responsibility over to `dftd4` to select the correct parameters or provides a complete damping parameters.

If Psi4 requires full control over the parameter selection, it can't use the former mechanism, but has to always construct the set of damping parameters first (`qcng` knows how to do this for all supported functionals in `dftd4`).

Note that in the future `dftd4` will most likely throw an error in case both functional name and an (incomplete) set of damping parameters are provided.

File: psi4/driver/procrouting/empirical_dispersion.py (hunk 7 above)
Comment (loriab) [reply psi4/driver/procrouting/empirical_dispersion.py:45]: yes, since black will clobber all my single quotes with double quotes someday, I've started to preempt it with doubles.

File: psi4/driver/qcdb/molecule.py (hunk 9 above)
Comment (loriab) [reply psi4/driver/qcdb/molecule.py]: I've revised the docstring, I hope truthfully. There were several principles that wanted to be followed here ((a) psi4 dft-d interface should be consistent, (b) qcengine harness should follow upstream project in keyword handling, (c) qcengine dash harnesses (dftd3, gcp, mp2d, dftd4) should be consistent, (d) Molecule.run_dftdN (aka psi4 -d interface) should be consistent), and I just couldn't make them all so. The `Molecule.run_dftdN` seemed like the least important one to sacrifice.

**Review by PeterKraus — APPROVED** [review]
Looks great, can't wait to use the functionals!
